File: C++/singly_linked_list.cpp
#include "singly_linked_list.h"

SinglyLinkedList::SinglyLinkedList() : head(nullptr), tail(nullptr), list_size(0) {}

SinglyLinkedList::~SinglyLinkedList() {
    clear();
}

void SinglyLinkedList::push_back(const string& value) {
    Node* new_node = new Node(value);
    if (!head) {
        head = tail = new_node;
    } else {
        tail->next = new_node;
        tail = new_node;
    }
    list_size++;
}

void SinglyLinkedList::push_front(const string& value) {
    Node* new_node = new Node(value);
    if (!head) {
        head = tail = new_node;
    } else {
        new_node->next = head;
        head = new_node;
    }
    list_size++;
}

void SinglyLinkedList::pop_front() {
    if (!head) return;
    Node* temp = head;
    head = head->next;
    delete temp;
    list_size--;
    if (!head) tail = nullptr;
}

void SinglyLinkedList::insert(size_t index, const string& value) {
    if (index > list_size) return;
    
    if (index == 0) {
        push_front(value);
        return;
    }
    if (index == list_size) {
        push_back(value);
        return;
    }

    Node* current = head;
    for (size_t i = 0; i < index - 1; ++i) {
        current = current->next;
    }
    
    Node* new_node = new Node(value);
    new_node->next = current->next;
    current->next = new_node;
    list_size++;
}

void SinglyLinkedList::remove(size_t index) {
    if (index >= list_size) return;
    
    if (index == 0) {
        pop_front();
        return;
    }

    Node* current = head;
    for (size_t i = 0; i < index - 1; ++i) {
        current = current->next;
    }
    
    Node* to_delete = current->next;
    current->next = to_delete->next;
    
    if (to_delete == tail) {
        tail = current;
    }
    
    delete to_delete;
    list_size--;
}

string& SinglyLinkedList::get(size_t index) {
    if (index >= list_size) throw out_of_range("Index out of range");
    
    Node* current = head;
    for (size_t i = 0; i < index; ++i) {
        current = current->next;
    }
    return current->data;
}

bool SinglyLinkedList::isEmpty() const {
    return list_size == 0;
}

size_t SinglyLinkedList::size() const {
    return list_size;
}

void SinglyLinkedList::clear() {
    while (!isEmpty()) {
        pop_front();
    }
}

vector<string> SinglyLinkedList::toVector() const {
    vector<string> result;
    Node* current = head;
    while (current) {
        result.push_back(current->data);
        current = current->next;
    }
    return result;
}

void SinglyLinkedList::fromVector(const vector<string>& elements) {
    clear();
    for (const auto& element : elements) {
        push_back(element);
    }
}

File: GO/SinglyLinkedList.go
package main

import "fmt"

type SLLNode struct {
	data string
	next *SLLNode
}

type SinglyLinkedList struct {
	head *SLLNode
	tail *SLLNode
	size int
}

func NewSinglyLinkedList() *SinglyLinkedList {
	return &SinglyLinkedList{}
}

func (s *SinglyLinkedList) PushBack(value string) {
	newNode := &SLLNode{data: value}
	if s.head == nil {
		s.head = newNode
		s.tail = newNode
	} else {
		s.tail.next = newNode
		s.tail = newNode
	}
	s.size++
}

func (s *SinglyLinkedList) PushFront(value string) {
	newNode := &SLLNode{data: value}
	if s.head == nil {
		s.head = newNode
		s.tail = newNode
	} else {
		newNode.next = s.head
		s.head = newNode
	}
	s.size++
}

func (s *SinglyLinkedList) PopFront() {
	if s.head == nil {
		return
	}
	s.head = s.head.next
	s.size--
	if s.head == nil {
		s.tail = nil
	}
}

func (s *SinglyLinkedList) Insert(index int, value string) {
	if index > s.size {
		return
	}
	if index == 0 {
		s.PushFront(value)
		return
	}
	if index == s.size {
		s.PushBack(value)
		return
	}
	
	current := s.head
	for i := 0; i < index-1; i++ {
		current = current.next
	}
	
	newNode := &SLLNode{data: value}
	newNode.next = current.next
	current.next = newNode
	s.size++
}

func (s *SinglyLinkedList) Remove(index int) {
	if index >= s.size {
		return
	}
	if index == 0 {
		s.PopFront()
		return
	}
	
	current := s.head
	for i := 0; i < index-1; i++ {
		current = current.next
	}
	
	current.next = current.next.next
	if current.next == nil {
		s.tail = current
	}
	s.size--
}

func (s *SinglyLinkedList) Get(index int) (string, error) {
	if index >= s.size {
		return "", fmt.Errorf("index out of range")
	}
	
	current := s.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	return current.data, nil
}

func (s *SinglyLinkedList) IsEmpty() bool {
	return s.size == 0
}

func (s *SinglyLinkedList) Size() int {
	return s.size
}

func (s *SinglyLinkedList) Clear() {
	for !s.IsEmpty() {
		s.PopFront()
	}
}

func (s *SinglyLinkedList) ToVector() []string {
	result := make([]string, 0, s.size)
	current := s.head
	for current != nil {
		result = append(result, current.data)
		current = current.next
	}
	return result
}

func (s *SinglyLinkedList) FromVector(elements []string) {
	s.Clear()
	for _, element := range elements {
		s.PushBack(element)
	}
}
